### ISSUE TITLE
dev-dependencies that are also extras

### ISSUE THREAD

**Issue by bertjwregeer** [python-poetry/poetry#129]
I've got a somewhat weird requirement:

Project A provides a bunch of pytest fixtures, and uses pytest itself for testing. I want to provide an extra on the resulting distributable that is called "testing".

So I created the following:

```
[tool.poetry.dependencies]
python = "*"
pyramid = "^1.9"
sqlalchemy = {version = "^1.2", extras = ["postgresql"]}
psycopg2 = {version = "^2.7", optional = true}
pytest = {version = "^3.5", optional = true}

[tool.poetry.dev-dependencies]
pytest = "^3.5"
pytest-cov = "^2.5"
coverage = "^4.5"

[tool.poetry.extras]
testing = ["pytest"]
postgresql = ["psycopg2"]
```

This will work correctly, however the dependency on `pytest` is specified twice. Once under `tool.poetry.dependencies` and another time under `tool.poetry.dev-dependencies`. What I would like to do is make sure that the same thing isn't listed twice, so I tried the following:

```
[tool.poetry.dependencies]
python = "*"
pyramid = "^1.9"
sqlalchemy = {version = "^1.2", extras = ["postgresql"]}
psycopg2 = {version = "^2.7", optional = true}

[tool.poetry.dev-dependencies]
pytest = "^3.5"
pytest-cov = "^2.5"
coverage = "^4.5"

[tool.poetry.extras]
testing = ["pytest"]
postgresql = ["psycopg2"]
```

Hoping that the `testing` extra would pick up `pytest` from `dev-dependencies`. However this is not the case.

So I moved `pytest` back up, and made it `optional`:

```
[tool.poetry.dependencies]
python = "*"
pyramid = "^1.9"
sqlalchemy = {version = "^1.2", extras = ["postgresql"]}
psycopg2 = {version = "^2.7", optional = true}
pytest = {version = "^3.5", optional = true}

[tool.poetry.dev-dependencies]
pytest-cov = "^2.5"
coverage = "^4.5"

[tool.poetry.extras]
testing = ["pytest"]
postgresql = ["psycopg2"]
```

Now it is correctly picked up, but there is no way to specify that when installing `dev-dependencies` that the extra `testing` should be automatically installed too, since that now contains the `pytest` dependency.

Trying:

```
mypackage = { version = "*", extras = ["testing"]}
```

Did not work since it could not find `mypackage`, and I am sure that if it could find it, it would likely download it from pypi which is not intended.

Is there a way to refer to `self` somehow? Or would it be possible to add a dependency in one place (define its version and extras/requirements) without redefining it in `dev-dependencies`?

The goal ultimately is that a secondary package would have the following `pyproject.toml`:

```
[tool.poetry]
name = "B-package"
version = "0.2.0"
description = ""
authors = ["Your Name <you@example.com>"]

[tool.poetry.dependencies]
python = "*"
mypackage = {version = "^0.1"}

[tool.poetry.dev-dependencies]
mypackage = {version = "*", extras = ["testing"]}
```

And only if you are developing would it install `mypackage[testing]` but for production it would install `mypackage`. Using the version constraint from `tool.poetry.dependencies` vs the constraint specified in `tool.poetry.dev-dependencies`.

This is related to secondary issue: https://github.com/sdispater/poetry/issues/128

**Comment by dmfigol** [comment]
I have similar use-cases for ReadTheDocs automatic builds: 
https://github.com/rtfd/readthedocs.org/issues/4912

**Comment by sobolevn** [comment]
I also have an optional dev-dependency in `wemake-django-template`: 

```
[tool.poetry.dev-dependencies]
ipython = {version = "*", optional = true }
```

Currently it does not work. https://github.com/wemake-services/wemake-django-template/issues/685

**Comment by chrahunt** [comment]
@sobolevn this is by design as mentioned in https://github.com/sdispater/poetry/pull/606#issuecomment-437943927.

**Comment by cjw296** [comment]
This is still valid.

**Comment by thejohnfreeman** [comment]
I believe the best solution to this has been proposed in #1644.

**Comment by finswimmer** [comment]
As @thejohnfreeman  wrote, this will be possible with dependency groups which are planned for poetry 1.2. I'll close this one in favor of #1644.

**Comment by sh-at-cs** [comment]
This hasn't really been solved by #1644, has it? It's still not possible to specify dependencies that belong to both an "extras" group *and* a Poetry group (like dev) without repeating them, version and all, is it? All that happened is that dev-dependencies became the "dev" group - how does that solve this?
